File: packages/pos-python-runtime/app/services/odoo_client.py
from typing import Any

import httpx


class OdooClientError(RuntimeError):
    pass


class OdooAuthError(OdooClientError):
    pass


class OdooClient:
    def __init__(self, client: httpx.AsyncClient):
        self._client = client

    async def login(self, login: str, password: str, device_code: str | None = None) -> dict[str, Any]:
        payload = {"login": login, "password": password, "device_code": device_code}
        return await self._post("/api/microservice/auth/login", payload)

    async def logout(self, odoo_access_token: str) -> dict[str, Any]:
        return await self._post(
            "/api/microservice/auth/logout",
            {},
            headers={"Authorization": f"Bearer {odoo_access_token}"},
        )

    async def _post(
        self,
        path: str,
        payload: dict[str, Any],
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        try:
            response = await self._client.post(path, json=payload, headers=headers)
        except httpx.HTTPError as exc:
            raise OdooClientError(f"Odoo request failed: {exc}") from exc

        if response.status_code >= 500:
            raise OdooClientError(f"Odoo HTTP {response.status_code}: {response.text[:300]}")

        try:
            data = response.json()
        except ValueError as exc:
            raise OdooClientError("Odoo returned non-JSON response.") from exc

        if isinstance(data, dict) and "result" in data and isinstance(data["result"], dict):
            data = data["result"]

        if response.status_code >= 400 or not data.get("success", False):
            message = data.get("message") or data.get("error") or "Odoo API rejected request."
            raise OdooAuthError(str(message))

        result = data.get("data", data)
        if not isinstance(result, dict):
            raise OdooClientError("Odoo API returned unexpected payload.")
        return result

File: services/product-catalog-service/src/odoo-client.ts
import { createHash } from "node:crypto";

import { AppConfig } from "./config.js";
import { upstreamError } from "./errors.js";

export type OdooCatalogRequest = {
  pos_config: number;
  offset: number;
  limit: number;
  updated_after?: string;
  barcode?: string;
  product_id?: number;
};

export async function fetchOdooCatalog(
  config: AppConfig,
  odooAccessToken: string,
  payload: OdooCatalogRequest
): Promise<Record<string, unknown>> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), config.requestTimeoutMs);
  try {
    const response = await fetch(`${config.odooBaseUrl}/api/microservice/catalog/products`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${odooAccessToken}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify(payload),
      signal: controller.signal
    });

    const body = await safeJson(response);
    const envelope = unwrapJsonRpc(body);
    if (!response.ok || envelope?.success !== true) {
      const message = envelope?.message || "Odoo catalog API rejected request.";
      throw upstreamError(String(message), envelope || body);
    }
    const data = envelope.data;
    if (!data || typeof data !== "object" || Array.isArray(data)) {
      throw upstreamError("Odoo catalog API returned invalid data payload.", envelope);
    }
    return data as Record<string, unknown>;
  } catch (error) {
    if (error instanceof Error && error.name === "AbortError") {
      throw upstreamError("Odoo catalog request timed out.");
    }
    throw error;
  } finally {
    clearTimeout(timeout);
  }
}

export type OdooProductImage = {
  contentType: string;
  data: Buffer;
  checksum: string;
  size: number;
  sourceUrl: string;
};

export async function fetchOdooProductImage(
  config: AppConfig,
  odooAccessToken: string,
  productId: number,
  sourceUrl?: string | null
): Promise<OdooProductImage | null> {
  const url = productImageUrl(config, productId, sourceUrl);
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), config.requestTimeoutMs);
  try {
    const response = await fetch(url, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${odooAccessToken}`,
        Accept: "image/avif,image/webp,image/apng,image/*,*/*;q=0.8"
      },
      signal: controller.signal
    });

    if (response.status === 404 || response.status === 204) return null;
    if (!response.ok) {
      throw upstreamError(`Odoo product image HTTP ${response.status}.`);
    }

    const contentType = (response.headers.get("content-type") || "application/octet-stream").split(";")[0].trim();
    if (!contentType.startsWith("image/")) return null;

    const contentLength = Number(response.headers.get("content-length") || 0);
    if (contentLength > config.catalogImageMaxBytes) {
      throw upstreamError(`Odoo product image is too large: ${contentLength} bytes.`);
    }

    const data = Buffer.from(await response.arrayBuffer());
    if (data.length === 0) return null;
    if (data.length > config.catalogImageMaxBytes) {
      throw upstreamError(`Odoo product image is too large: ${data.length} bytes.`);
    }

    return {
      contentType,
      data,
      checksum: createHash("sha256").update(data).digest("hex"),
      size: data.length,
      sourceUrl: url
    };
  } catch (error) {
    if (error instanceof Error && error.name === "AbortError") {
      throw upstreamError("Odoo product image request timed out.");
    }
    throw error;
  } finally {
    clearTimeout(timeout);
  }
}

function productImageUrl(config: AppConfig, productId: number, sourceUrl?: string | null): string {
  const base = new URL(config.odooBaseUrl);
  if (!sourceUrl) {
    return new URL(`/web/image/product.product/${productId}/${config.catalogImageField}`, base).toString();
  }

  try {
    const parsed = new URL(sourceUrl, base);
    const pathname = parsed.pathname.replace(/\/image(?:_\d+)?$/, `/${config.catalogImageField}`);
    return new URL(`${pathname}${parsed.search}`, base).toString();
  } catch {
    return new URL(`/web/image/product.product/${productId}/${config.catalogImageField}`, base).toString();
  }
}

function unwrapJsonRpc(body: unknown): Record<string, unknown> | undefined {
  if (!body || typeof body !== "object" || Array.isArray(body)) return undefined;
  const data = body as Record<string, unknown>;
  const result = data.result;
  if (result && typeof result === "object" && !Array.isArray(result)) {
    return result as Record<string, unknown>;
  }
  return data;
}

async function safeJson(response: Response): Promise<unknown> {
  try {
    return await response.json();
  } catch {
    return await response.text();
  }
}
